### DISCUSSION TITLE
Some black pictures on html

### DISCUSSION THREAD

**Discussion by Yichen-fqyd** [pymupdf/PyMuPDF#976] [Looking for help] [answered]
Hi, when I was converting pdf file to html file, sometimes there appears a total black image shown on html, while original pdf didn't show, 
What is the reason for this issue, and how to possibly solve it?
is there a way to get information from block==2 get("dict") to know that it is a totally black image?
{'bbox': (386.1600036621094, 82.44001007080078, 491.6400146484375, 103.32000732421875), 'bpc': 8, 'colorspace': 3, 'ext': 'png', 'height': 57, 'image': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x01%...\x00\x00IEND\xaeB`\x82', 'number': 10, 'type': 1, 'width': 293, 'xres': 96, 'yres': 96}

**Comment by JorjMcKie** [comment, marked as answer]
This a known bug - see #657. The image has an alpha channel. You can find this out by inspecting the list `page.get_images()`. For each image on the page, there is an item like ``(xref, smask, ...)`` in this list. If smask is > 0, then this image may show this unpleasant appearance in HTML output.

**Reply by Yichen-fqyd** [reply]
When I was trying this method, 
I found there is some discrepancies between the method page.get_images() vs imgblocks = [b for b in blocks if b["type"] == 1], page.get_images() sometimes return a list of one image, but imgblocks have length of three.

What is the difference between these two methods in extracting images from fitz_page?? and why the length is different?

Thank you

**Reply by JorjMcKie** [reply]
> What is the difference between these two methods in extracting images from fitz_page?? and why the length is different?

It is worth reading [this](https://pymupdf.readthedocs.io/en/latest/textpage.html#dictionary-structure-of-extractdict-and-extractrawdict) section of the documentation on this. Here is the most direct quote to your question:
```
TextPage and corresponding method Page.get_text() are available for all document types.
Only for PDF documents, methods Document.get_page_images() / Page.get_images() offer some
overlapping functionality as far as image lists are concerned. But both lists may or may not contain
the same items. Any differences are most probably caused by one of the following:

        “Inline” images (see page 352 of the Adobe PDF References) of a PDF page are contained in a textpage,
        but not in Page.get_images().

        Image blocks in a textpage are generated for every image location – whether or not there are any
        duplicates. This is in contrast to Page.get_images(), which will contain each image only once.

        Images mentioned in the page’s object definition will always appear in Page.get_images(). But it may
        happen, that there is no “display” command in the page’s contents (erroneously or on purpose). In this
        case the image will not appear in the textpage.
```

**Reply by JorjMcKie** [reply]
As a consequence of the post before: your PDF page must contain 2 inline images ...

**Reply by JorjMcKie** [reply]
**_Inline_** images of PDF pages typically are small images. They are contained in one of the page's `/Contents` objects. The relevant syntax can be found in the Adobe manual https://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/pdf_reference_1-7.pdf on page 985, "Operator Summary". Those images are enclosed in the string pair ``"BI"`` / ``"EI"``.

**Reply by Yichen-fqyd** [reply]
Thank you so much for all these detailed explanations!!!!
I will read through the documents you mentioned above.

another question, I found https://pymupdf.readthedocs.io/en/latest/document.html#Document.convert_to_pdf
Does this mean this will also support docx/doc file to pdf conversation?

Thanks again

**Reply by JorjMcKie** [reply]
>  Does this mean this will also support docx/doc file to pdf conversation?

No, conversion only works, if the input is an MuPDF-supported document type. ``docx/doc`` (presumably a Word document) is not among these types.
But you do have the PDF export functions of either MS Office or LibreOffice, etc.
Admittedly, this is not working inside Python scripts. You could figure out, how a batch command for a word->PDF conversion looks like (I am sure you will find something, or write a Word macro yourself), then invoke this via Python's ``subprocess`` and read the produced output.

**Reply by Yichen-fqyd** [reply]
Got it!! Thank you for the advice, will try this method!!

**Reply by Yichen-fqyd** [reply]
Hi, I repeated read the part where three conditions why these two methods are different, I am wondering for the second term: 

> Image blocks in a textpage are generated for every image location – whether or not there are any
> duplicates. This is in contrast to Page.get_images(), which will contain each image only once.

I am not quite sure I understand this, when would image blocks have duplicates? Do you mean two images on the same location? 

Thank you so much

**Reply by JorjMcKie** [reply]
No - just the other way round: the same image in two locations, like here:
[imgfile3.pdf](https://github.com/pymupdf/PyMuPDF/files/6281045/imgfile3.pdf)
`doc.get_page_images(0)` will report just one image, although 2 are shown with the same content (and pointing to the same xref). And `page.get_text("dict")` will report both of them (with identical contents, but different bboxes).
Everything perfectly legal and by the books ...

**Reply by JorjMcKie** [reply]
The coming new version 1.18.11 will contain a new `Page` method `get_image_info()`. It reports the same content as the `get_text("dict")` method for images. But it will not return text and not the binary image content - only the respective images sizes.
In addition, there will be thea new dictionary item ``transform``, which will contain information about what happened to the original image to make it fit into that page's bbox - IAW scaling or rotation can then be detected in detail.

**Reply by Yichen-fqyd** [reply]
Thank you for your detailed explanation, now I am clear.


Looking forward the new version 1.18.11, I am pretty sure this function `get_image_info()` will be very helpful!!!

**Reply by Yichen-fqyd** [reply]
> As a consequence of the post before: your PDF page must contain 2 inline images ...



Hi,
Follow up for the inline images, if the inline images only present in the image blocks, how can I get information whether the image has mask or alpha channel,

Thank you

**Reply by JorjMcKie** [reply]
There currently is no way to determine whether an *inline* image has an alpha channel - and consequently there is no way to access the alpha information.
For other images, this can be found out by inspecting the items of list `page.get_images()`. You can use `page.get_image_info(xref=True)` which is a list of image meta info dictionaries displayed by the page. Those dictionaries contain an "xref" key, which points to a non-inline image if >0.

**Reply by JorjMcKie** [reply]
Additional note to my previous post:
I have never checked this out: MuPDF detects inline images and returns the corresponding binary as part of the image block in `page.get_text("dict")`. There is a chance, that this logic is smart enough to include any existing alpha channel. Can be checked by making a pixmap of that bytes object and look at pix.alpha.
